Fix missing parentheses in LaTex and pprint output

Subhash-Saurabh opened the pull request:
Missing Parentheses in Latex Output and pprint output #13651 

 The earlier code was giving error in LaTeX  and pprint representations of sums containing 
 sums multiplied by -1.
previous -
```
>>> import sympy
>>> from sympy.abc import a, b, c
>>> buggy = c + sympy.collect_const(-a - b)
>>> sympy.latex(buggy)
>>> sympy.pprint(buggy)
```

Latex Output ->  'c - a + b
pprint Output -> c - a + b    

now -
```
>>> import sympy
>>> from sympy.abc import a, b, c
>>> buggy = c + sympy.collect_const(-a - b)
>>> sympy.latex(buggy)
>>> sympy.pprint(buggy)
```
Latex Output ->   c - \left(a + b\right)
pprint Output -> c - (a + b)

Subhash-Saurabh committed:
issue_13651

gxyd commented:
Please read this https://github.com/sympy/sympy/wiki/Development-workflow#title-summary and https://github.com/sympy/sympy/wiki/Development-workflow#writing-commit-messages

You can use triple quotes (I did for now) to better format the code in the description https://guides.github.com/features/mastering-markdown/

Subhash-Saurabh committed:
issue_13651 tests

asmeurer commented on sympy/printing/latex.py:
```diff
@@ -300,7 +300,7 @@ def _print_Add(self, expr, order=None):
                 tex += " + "
             term_tex = self._print(term)
-            if self._needs_add_brackets(term):
+            if self._needs_add_brackets(term) or term.is_Add:
```
It seems this check should be in _needs_add_brackets itself.

Subhash-Saurabh committed:
Missing bracket in latex output #13651

Subhash-Saurabh commented:
@asmeurer  I have made the changes you suggested .Can you please review it now.

gxyd commented on sympy/printing/tests/test_latex.py:
```diff
@@ -1630,6 +1631,10 @@ def test_issue_12886():
     assert latex(m__1**2 + l__1**2) == r'\left(l^{1}\right)^{2} + \left(m^{1}\right)^{2}'
 
+def test_issue_13651():
+    from sympy.abc import a,b,c
```
Symbols 'a, 'b' and 'c' are defined at the top of file.

gxyd commented on sympy/printing/tests/test_latex.py:
```diff
@@ -1630,6 +1631,10 @@ def test_issue_12886():
     m__1, l__1 = symbols('m__1, l__1')
     assert latex(m__1**2 + l__1**2) == r'\left(l^{1}\right)^{2} + \left(m^{1}\right)^{2}'
 
```
Add one more empty new line here for consistency.

gxyd commented on sympy/printing/tests/test_latex.py:
```diff
@@ -1630,6 +1631,10 @@ def test_issue_12886():
 
+def test_issue_13651():
+    from sympy.abc import a,b,c
+    bug = c+collect_const(-a-b)
```
`bug` isn't a good variable name. Also use space `-a - b` and `c + ...`

gxyd commented on sympy/printing/tests/test_latex.py:
```diff
@@ -1630,6 +1631,10 @@ def test_issue_12886():
+    from sympy.abc import a,b,c
+    bug = c+collect_const(-a-b)
+    assert latex(bug) == r"c - \left(a + b\right)"
 
```
Add one more empty new line here for consistency.

gxyd commented:
Also address this comment

https://github.com/sympy/sympy/pull/13673#issuecomment-349654990

Subhash-Saurabh committed:
Requested changes in test_latex

printing/tests/test_latex : Made requsted changes for code consistency

Subhash-Saurabh committed:
Merge branch 'master' into 13651_issue

gxyd commented:
On your branch:
```python
>>> import sympy
>>> from sympy.abc import a, b, c
>>> buggy = c + sympy.collect_const(-a - b)
>>> sympy.pprint(buggy)
c - a + b
```

Subhash-Saurabh commented:
@gxyd I have only modified the code for latex part . I am currently working on the pprint bug and I will commit as soon as possible.

Subhash-Saurabh commented:
@gxyd For the pprint part should I submit a new PR or make additional commits in this PR ?

gxyd commented:
It will be preferable if you fix that in this PR. Since, I think it is related to making the same expression work for different printings.

Subhash-Saurabh committed:
Corrected Missing Parentheses in pprint output

printing/pretty/pretty.py: Added a extra check to check if parentheses are
required or not
printing/pretty/test_pretty.py : Created additional tests

Subhash-Saurabh committed:
Merge branch '13651_issue' of https://github.com/Subhash-Saurabh/sympy into
13651_issue

Subhash-Saurabh committed:
Minor fixes

Removed Trailing whitespace in test_latex.py

Subhash-Saurabh commented:
@gxyd I have added the fix for pprint part also. Can you please review it and suggest if any changes are to be made.

gxyd commented on sympy/printing/pretty/pretty.py:
```diff
@@ -1416,7 +1416,7 @@ def pretty_negative(pform, index):
                 pform_neg = ' - '
 
-            if pform.binding > prettyForm.NEG:
+            if pform.binding > prettyForm.NEG or pform.binding == prettyForm.ADD:
```
Can you break this line using `( )` brackets around the complete line, so that line length is less than 79 characters?

gxyd commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -7,7 +7,7 @@
     groebner, oo, pi, symbols, ilex, grlex, Range, Contains,
     SeqPer, SeqFormula, SeqAdd, SeqMul, fourier_series, fps, ITE,
-    Complement, Interval, Intersection, Union, EulerGamma, GoldenRatio)
+    Complement, Interval, Intersection, Union, EulerGamma, GoldenRatio, collect_const)
```
put `collect_const` on the next line.

gxyd commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -5849,6 +5849,13 @@ def test_issue_9877():
 
 
+def test_issue_13651():
+    expr = c + collect_const(-a - b)
```
`expr` -> `expr1`.

gxyd replied:
instead have `expr1 = c + Mul(-1, a + b, evaluate=False)` not `collect_const`.

gxyd commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -5849,6 +5849,13 @@ def test_issue_9877():
+def test_issue_13651():
+    expr = c + collect_const(-a - b)
+    assert pretty(expr) == 'c - (a + b)'
+    expr = c + collect_const(-a + b - d)
```
`expr` -> `expr2`.

gxyd replied:
Also use `Mul` here as well.

gxyd commented on sympy/printing/tests/test_latex.py:
```diff
@@ -1641,6 +1642,11 @@ def test_issue_13559():
 
 
+def test_issue_13651():
+    expr = c + collect_const(-a - b)
```
use `Mul(-1, a + b, evaluate=...)` just like above comment.

gxyd reviewed (commented):
instead of `collect_const` use `Mul(-1, ..., evaluate=False)`, I believe that will be a more direct test. See this comment of Aaron https://github.com/sympy/sympy/issues/13651#issuecomment-349149751

Apart from these comments, this PR looks good to me.

Subhash-Saurabh committed:
Requested changes in tests

test_latex.py: Instead of collect_const used Mul
test_pretty.py: changes similar to test_latex
pretty.py: Minor change to keep length of line small

gxyd reviewed (approved):
LGTM for merging, once tests pass.

gxyd commented:
Thanks for your contribution. Good work.

gxyd commented:
Also make sure you use 'Fixes #NNNN' in one of your commits, *not* just #NNNN. This way the issue will be automatically closed.
